### ISSUE TITLE
Distinguishing capital and lower case letters in key bindings

### ISSUE THREAD

**Issue by lehmacdj** [next-browser/next#58]
I took a stab at implementing some vi keybindings, however encountered a problem while doing so. Unfortunately capital and lower case letters are recognized as the same keystroke. Consider the following simple `init.lisp` as an example:

```
(in-package :next)
(define-key *global-map* (kbd "j") #'scroll-down)
(define-key *global-map* (kbd "J") #'scroll-to-top)
```

Regardless of whether "J" or "j" is pressed the window always scrolls to the top instead of differentiating its behavior as necessary. I think this is because the way key strokes are reported, "J" and "j" produce the same key code just with the shift modifier either set or unset. It would be ideal if we could allow using the different letters for defining keybindings, and differentiate them however necessary.

I can try to implement the ability to differentiate these in the next day or two.

**Comment by jmercouris** [comment]
Hi, thank you for the report! I did not know about this.

I think I know actually what may be the cause of this. If you take a look at `keymap.lisp` and look at `push-key-chord`, you'll see the function you need to examine. There is where you can test to see whether the system is actually pushing lowercase and uppercase letters. If it is pushing upper and lower case letters correctly, then the problem must lie within the consumption of key-chords or within the key-chord struct itself.

For this, I would take a look at `consume-key-sequence` also within `keymap.lisp`. Possibly the equality function being used for the keymap is too liberal when comparing the key-chord structs. You can see the clhs entry here: http://www.lispworks.com/documentation/HyperSpec/Body/f_mk_has.htm

make-hash accepts a test parameter, in this case I have set it to `'equalp` for most of the key mappings within next. You can see an example of this in `global.lisp`, on the line that starts with `defvar *global-map*`. Therefore perhaps making a keymap with a different equality comparator might fix the issue.

There are a couple of equality comparators that may be more suitable. A list of the types of equalities can be seen here: http://www.gigamonkeys.com/book/syntax-and-semantics.html in the section titled `Truth, Falsehood, and Equality`

Once again, thank you for your report, please let me know how the investigation goes!

**Comment by lehmacdj** [comment]
The problem is that structs cannot be compared with `'equal`, but `'equalp' uses a case insensitive comparison. The keys are being reported with the correct case.

I think the correct solution is to use a custom comparison function, for which we would also need a custom hash function, or as a simpler hack, instead of using a struct we could use a list of a character followed by 3 booleans to represent the key struct and then use `'equal`, but that seems bad from a readability perspective.

**Comment by lehmacdj** [comment]
I hacked together a fix for this, but the code quality is terrible and I am not at all happy with the solution I came up with. 

You can find my solution [here](https://github.com/lehmacdj/nEXT/blob/keymap-case-sensitive/next/source/keymap.lisp). Essentially what I do is convert a `key` struct into a list of its components before using it as a key of the hash map. Then it is possible to use `'equal` instead of `'equalp` because `'equal` compares lists element wise.

 I also looked into implementing a separate hash function and equality for `key` structs but it turns out that using different equalities and hash functions is not standardized in common lisp, thus would require another library. Ultimately that may end up being the cleanest solution though.

**Comment by jmercouris** [comment]
Hmm yeah, I saw the solution, you are right, we need to come up with something better. It's not ideal, but probably better than pulling in another dependency.


Let me think about this overnight and see if anything comes to mind. I'll try to ask as well tomorrow on #lisp, maybe someone has some ideas.

**Comment by jmercouris** [comment]
I did actually think of another solution, but also not an optimal one. We can add a field to the struct titled `case-upper-p`.

Another option is to have have the `shift-modifier` be represented in the struct. When adding key binds with an upper-case letter, set the `shift-modifier` to be `T`. When pushing a key-chord with a capital letter also set `shift-modifier` to be `T`.

I haven't asked on #lisp yet, but I will be sure to do so, as none of these solutions are particularly elegant or clean :D

The shift modifier would work okay with something like `s+j`, but it would not be able to differentiate between `c-g` and `c-G`. I'm not sure how important that is, but just something to consider.

**Comment by lehmacdj** [comment]
That is definitely the most reasonable solution so far. 

I don’t exactly understand why we wouldn’t be able to distinguish c-g and c-G but I don’t think that is a very big problem. It would be nice to have but isn’t super important in my opinion.

**Comment by jmercouris** [comment]
The reason we would not be able to distinguish c-g and c-G would be in the event that the user has the caps lock key enabled for the pressing of g in c-G, this would create a capital G with no shift modifier

**Comment by lehmacdj** [comment]
Ah, I think just checking the key we get is better then maybe? But perhaps we want to be able to bind shift-x and caps lock x? I have caps lock rebound to control thus don’t really have much stake in that particular issue.

**Comment by jmercouris** [comment]
I have great news! Here's what we can do:

1. convert string to char in struct key - (key-character) is actually technically a string
2. convert char to code using (char-code)

repl:

```
CL-USER> (char-code #\a)
97
CL-USER> (char-code #\A)
65
CL-USER> (code-char 65)
#\A
```

chat transcript:

```
> I have a hash-table that uses a struct as a key. I use equalp
    comparison as the test for my hash table. My struct is composed of
    a field "character" which has a value of a character that may be
    upper or lowercase. Currently, my hash-table cannot distinguish
    between an upper and lower case character, how can I fix this?
<loke> jmercouris: Yes. You have discovered one of the worst parts of
    CL.
<loke> jmercouris: What you want to do can't be done with hashtables
    as defined by the spec. Most CL implementations however implement
    support for custom hashes. You can use those.
<loke> jmercouris: Or, you can use an alternative hashtable
    implementations. There are several on QL.
<rme> can't you use an equal hash table instead of an equalp hash
    table?
<loke> rme: EQUAL on structs acts as EQ
<loke> I.e. identity comparison
<loke> http://clhs.lisp.se/Body/f_equal.htm
> loke: That's too bad
> I was hoping to avoid something like that, but it may be necessary
    it seems
<loke> jmercouris: If you're OK with sticking to a given
    implementation of CL, you can always use vendor extensions
<loke> Otherwise, the third-party implementations are good.
> I have to support sbcl and ccl, so that adds some complexity
<loke> I think cl-containers has one
> I may have to use a third party implementation, or add a struct slot
    "uppercase-p" or something
<loke> jmercouris: You'd need one mer character. :-)
<loke> per
<loke> You could also save the strings as UTF-8 encoded byte arrays
<loke> (horrific, I know)
> loke: I'd need one more character? what do you mean?
<loke> jmercouris: You' dneed to store the upper/lowercase state for
    each character.
<loke> You can't distinguish between FOO Foo and foo using only one
    bit.
> loke: My struct only supports a single char
<loke> I see
> so I can distinguish between a and A using only one bit
<loke> jmercouris: Store it as a numeric code (CHAR-CODE) then.
<loke> The thing is, upper/lowercase is actually a uch more
    complicated concept than you might think, and the CL spec is so
    weak on the subject that you can't rely on it.
<loke> (mainly because CL was designed before Unicode)
> I think char-code is good actually
> I'll actually truly have to convert to a char, then a char code, but
    that makes sense
> it is actually a string of length 1 currently, but char-code it is
<rme> loke: thanks; you're quite right
> what's the reverse of char-code?
> (code-char?)
> wow, it actually is that
> thank you for your help
```

**Comment by lehmacdj** [comment]
closed since #62 was merged.
